src/gsi/stppm2_5.f90

```diff
@@ -161,6 +166,100 @@ subroutine stppm2_5(pm2_5head,rval,sval,out,sges,nstep)
+                j1=pm2_5ptr%ij(1)
+                j2=pm2_5ptr%ij(2)
+                j3=pm2_5ptr%ij(3)
+                j4=pm2_5ptr%ij(4)
+                j5=pm2_5ptr%ij(5)
+                j6=pm2_5ptr%ij(6)
+                j7=pm2_5ptr%ij(7)
+                j8=pm2_5ptr%ij(8)
+                w1=pm2_5ptr%wij(1)
+                w2=pm2_5ptr%wij(2)
+                w3=pm2_5ptr%wij(3)
+                w4=pm2_5ptr%wij(4)
+                w5=pm2_5ptr%wij(5)
+                w6=pm2_5ptr%wij(6)
+                w7=pm2_5ptr%wij(7)
+                w8=pm2_5ptr%wij(8)
+
+                istatus=0
+
+                val2=-pm2_5ptr%res
+                val=zero
+
+                !aeroname='aso4i'
+          !!!
+          do iaero=1,naero_cmaq_fv3
+          aeroname=aeronames_cmaq_fv3(iaero)
+
+                call gsi_bundlegetpointer(sval,trim(aeroname),spm2_5,istatus)
+                if(istatus /= 0) then
+                   write(6,*) 'error gsi_bundlegetpointer in stppm2_5 for ',&
+                        aeroname
+                   call stop2(456)
+                endif
+
+                call gsi_bundlegetpointer(rval,trim(aeroname),rpm2_5,istatus)
+
+                if(istatus /= 0) then
+                   write(6,*) 'error gsi_bundlegetpointer in stppm2_5 for ',&
+                        aeroname
+                   call stop2(457)
+                endif
+
+                !this doesn't follow the logic of wrfchem 
+                val= pm2_5ptr%pm25wc(imodes_cmaq_fv3(iaero))* & 
+                    (w1* rpm2_5(j1)+w2* rpm2_5(j2)+w3* rpm2_5(j3)+w4*rpm2_5(j4)+ &
+                     w5* rpm2_5(j5)+w6* rpm2_5(j6)+w7* rpm2_5(j7)+w8*rpm2_5(j8))+val
+                val2=pm2_5ptr%pm25wc(imodes_cmaq_fv3(iaero))* &
+                    (w1* spm2_5(j1)+w2* spm2_5(j2)+w3* spm2_5(j3)+w4*spm2_5(j4)+ &
+                     w5* spm2_5(j5)+w6* spm2_5(j6)+w7* spm2_5(j7)+w8*spm2_5(j8))+val2
+
+
+                nullify(spm2_5,rpm2_5)
+           end do
+                do kk=1,nstep
+                   qq=val2+sges(kk)*val
+                   pen(kk)=qq*qq*pm2_5ptr%err2
+                end do
+             else
+                pen(1)=pm2_5ptr%res*pm2_5ptr%res*pm2_5ptr%err2
+             end if
+
+!  modify penalty term if nonlinear qc
+
+             if (nlnqc_iter .and. pm2_5ptr%pg > tiny_r_kind .and. &
+                  pm2_5ptr%b  > tiny_r_kind) then
+                pm2_5_pg=pm2_5ptr%pg*varqc_iter
+                cg_pm2_5=cg_term/pm2_5ptr%b
+                wnotgross= one-pm2_5_pg
+                wgross = pm2_5_pg*cg_pm2_5/wnotgross
+                do kk=1,max(1,nstep)
+                   pen(kk)= -two*log((exp(-half*pen(kk))+wgross)/(one+wgross))
+                end do
+             endif
+
+             out(1) = out(1)+pen(1)*pm2_5ptr%raterr2
+             do kk=2,nstep
+                out(kk) = out(kk)+(pen(kk)-pen(1))*pm2_5ptr%raterr2
+             end do
+          end if
+
+          !pm2_5ptr => pm2_5ptr%llpoint
```

Comment:
I think I already mentioned something similar earlier, but remove this if not needed.